compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -13570,10 +13695,37 @@ TR::Node *ifCmpWithEqualitySimplifier(TR::Node * node, TR::Block * block, TR::Si
 
    makeConstantTheRightChildAndSetOpcode(node, firstChild, secondChild, s);
 
+   TR::ILOpCodes opCode = node->getOpCodeValue();
+
+   // Reduce ifcmpge to ifcmpeq in specific cases
+   IfcmpgeReducer ifcmpReducer(node, s);
+   switch (opCode)
+      {
+         case TR::ifbcmpge:
+            if (ifcmpReducer.reducible<int8_t>())
+               node = ifcmpReducer.reduce();
+            break;
+
+         case TR::ifbucmpge:
+            if (ifcmpReducer.reducible<uint8_t>())
+               node = ifcmpReducer.reduce();
+            break;
+
+         case TR::ifscmpge:
+            if (ifcmpReducer.reducible<int16_t>())
+               node = ifcmpReducer.reduce();
+            break;
+
+         case TR::ifsucmpge:
+            if (ifcmpReducer.reducible<uint16_t>())
+               node = ifcmpReducer.reduce();
+            break;
+      }
```

Comment:
If you wrap `reducible<T>` as I suggested above, this switch would simplify into something like this:

    TR::ILOpCode op = node->getOpCode();
    if (op.isCompareForOrder() && op.isCompareTrueIfGreater() && ifcmpReducer.reducible())
       node = ifcmpReducer.reduce();

Note that `op.isCompareTrueIfEqual()` is guaranteed by the choice of opcodes that use `ifCmpWithEqualitySimplifier`, and it should be preserved by `makeConstantTheRightChildAndSetOpcode` (through `getOpCodeForSwapChildren`).

Reply:
Fixed.